### ISSUE TITLE
hessian_vector_product doesn't work for matrix arguments

### ISSUE THREAD

**Issue by j-towns** [HIPS/autograd#86]
I'm working on a [toolbox](https://github.com/j-towns/pymanopt) which uses autograd as an autodiff backend. We need to calculate hessian-vector products when the 'vector' is actually a matrix, along the lines of Theano's Rop.

I'm wondering whether you'd like to include this functionality in autograd. I'm actually quite interested in how this works and could have a go at implementing it myself and submit a pr. Let me know your thoughts.

**Comment by mattjj** [comment]
Hey, glad you're using autograd!

Actually, Theano's Rop doesn't compute a Hessian-vector product, but rather [it computes a Jacobian-vector product](http://deeplearning.net/software/theano/tutorial/gradients.html#r-operator). That is, it only uses first-derivative information. (Theano's Lop is pretty much the same operation that autograd's gradfuns apply to derivatives during its evaluation of the chain rule, and as such Lop can be used in an autograd primitive to [make Theano handle some gradient calculations on the GPU](https://github.com/andymiller/theanograd/blob/a0704af8c27a782ab11adc8eeb7dc9041407efd2/theanograd/mlp.py#L56).)

The [`hessian_vector_product` code](https://github.com/HIPS/autograd/blob/eff866a5afde4cf351e923dc990402e72a206079/autograd/convenience_wrappers.py#L66-L74) is actually quite short, and the only real computation is done by `np.dot`. It might be that by replacing `np.dot` with another product, like one based on `einsum`, could make your use case work, or maybe a call to `grad` needs to be replaced with a call to `jacobian`. However, I'm not sure exactly what computation you need or where the current implementation breaks.

Can you post example code, including the input you're using and the output you expect?

**Comment by mattjj** [comment]
Looking at the code a bit more, I suspect we just need to change the second `grad` call to a `jacobian` call. That is, the output of `vector_dot_grad` won't be a scalar if the "vector" is a matrix, and so `grad` isn't the right derivative function to use.

**Comment by j-towns** [comment]
Ah thank you for getting back so quickly. Currently we are using the following code to compute the thing we need. `objective` is a function which takes a tensor (either a vector, a matrix or a higher order tensor) as an argument and returns a scalar.

```
def hessian(objective):
     def hess(x, g):
         return np.tensordot(ad.hessian(objective)(x), g, axes=x.ndim)
     return hess
```

The above is very slow since the computation required is quadratic in the size of x and g. I was hoping the following would work, and produce the same thing:

```
def hessian(objective):
    def hess(x, g):
        return hessian_vector_product(objective)(x, g)
    return hess
```
As you know it does work, but only if x and g are 1d arrays.

So do you think this would be a straightforward fix?

**Comment by j-towns** [comment]
One reason this makes sense is because the hessian vector product is the directional derivative of the gradient in the direction of the vector. What we need is precisely a higher dimensional analogue to this.

**Comment by mattjj** [comment]
Ah I see, thanks for explaining, that's interesting.

I just pushed a branch which makes the change I mentioned (switch a `grad` to a `jacobian`) and adds two tests. See [this Hessian-matrix product test](https://github.com/HIPS/autograd/blob/7ea4a7168d25d72adde33323126c6ad4d4cbd4b3/tests/test_wrappers.py#L85-L90) to see how it's meant to work. Note the transpose!

Can you try out the branch and see how things look?

**Comment by j-towns** [comment]
Thanks so much for the help. Unfortunately that wasn't quite what I wanted, though similar. I've submitted a pull request which is functioning how I would hope (essentially just changed the dot to a tensordot).

**Comment by mattjj** [comment]
Awesome, thanks! (Commented on the PR as well.)

I'll leave this issue open until we merge into master in the very near future.

**Comment by mattjj** [comment]
I think we closed this with PR #88.
